### PR TITLE
JBIDE-23505 - initial impl of cdkv3 server adapter / minishift based

### PR DESCRIPTION
**Pull request by robstryker**
Not for commit, just here for discussion, testing, and critique.  Repeat: DO NOT COMMIT

# Pull Request Checklist
## General
* Is this a blocking issue or new feature? If yes, QE needs to +1 this PR

New feature

## Code
* Are method-/class-/variable-names meaningful?    Yes
* Are methods concise, not too long?   Yes
* Are catch blocks catching precise Exceptions only (no catch all)?   Yes

## Testing
* Are there unit-tests?   No
* Are there integration tests (or at least a jira to tackle these)? No
* Is the non-happy path working, too?  Yes
* Are other parts that use the same component still working fine?   Need to check

## Function
* Does it work?   I think so

### CODE REVIEW

**Comment by adietish** [comment]
@robstryker @jeffmaury I tend to think that we want this to get in for 4.5.0.AM1, not 4.4.2.Final. Agree?

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/MinishiftBinaryUtility.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.wst.server.core.IServer;
+import org.jboss.tools.openshift.cdk.server.core.internal.adapter.controllers.IExternalLaunchConstants;
+import org.jboss.tools.openshift.internal.common.core.util.CommandLocationBinary;
+
+public class MinishiftBinaryUtility {
+	private static final String VAGRANT_LOCATION_LINUX = "/usr/bin/minishift";
+	
+	// Seems weird?  See https://github.com/mitchellh/vagrant/issues/1652
+	private static final String VAGRANT_LOCATION_WINDOWS = "C:\\minishift.exe";
```
Comment (jeffmaury) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/MinishiftBinaryUtility.java]: Constant should be renamed

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/MinishiftBinaryUtility.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
```
Comment (jeffmaury) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/MinishiftBinaryUtility.java]: I think code can be shared/refactored with VagrantBinaryUtility

File: plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/adapter/MinishiftPoller.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	
+	private File getWorkingDirectory(IServer s) throws PollingException {
+		return JBossServerCorePlugin.getServerStateLocation(s).toFile();
+	}
+		
+	
+	protected IStatus onePing(IServer server, Map<String, String> env)
+			throws PollingException, IOException, TimeoutException {
+
+		String[] args = new String[] { CDKConstants.VAGRANT_CMD_STATUS };
```
Comment (jeffmaury) [inline plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/adapter/MinishiftPoller.java]: I think there is lots of code refereing to Vagrant concept/naming that need to be renamed in order for this code to be readable and maintainable

**Comment by adietish** [comment]
@robstryker can you please rebase this PR?

**Comment by adietish** [comment]
steps:
1. download and install the latest alpha2 minishift binary (http://cdk-builds.usersys.redhat.com/builds/weekly/25-Jan-2017.3.0_alpha.2)
2. configure the adapter to use this
3. create the adapter
4. start the adapter

Result:
It wont recognize the result of the latest minishift binary. The start actually succeeded, but the adapter failed to recognize it, reported that starting failed.
image

**Comment by adietish** [comment]
What's also not working as far as I can see, is the ability to detect a running minishift (ex. you launch minishift on cli and then start the adapter)

**Comment by adietish** [comment]
This server adapter also seems to miss a server type. 
Once you have this server adapter created and you try to create an OpenShift server adapter you'll fail to do so. The server adapter creating code will NPE where it tries to see if we already have an adapter for the given OpenShift service. The NPE occurs where the server type of the CDK3 server adapter inspected and compared to the OpenShift server type:

https://github.com/jbosstools/jbosstools-openshift/blob/master/plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/OpenShiftServerUtils.java#L98
```
Stream.of(ServerCore.getServers())
				.filter(server -> server.getServerType()
						.equals(serverType)
``` 

I ran into the following exception when the above code ran into this CDK/minishift server adapter:
```
org.eclipse.e4.core.di.InjectionException: java.lang.NullPointerException
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:65)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:282)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:264)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:132)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:152)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:494)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:488)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:210)
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.executeItem(HandledContributionItem.java:433)
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem.handleWidgetSelection(AbstractContributionItem.java:454)
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem$3.handleEvent(AbstractContributionItem.java:482)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4256)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1501)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1509)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1313)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4080)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3706)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1121)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1022)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:150)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:693)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:610)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:138)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:673)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:610)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1519)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1492)
Caused by: java.lang.NullPointerException
	at org.jboss.tools.openshift.core.server.OpenShiftServerUtils.lambda$1(OpenShiftServerUtils.java:102)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
	at java.util.stream.ReferencePipeline$11$1.accept(ReferencePipeline.java:373)
	at java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
	at org.jboss.tools.openshift.core.server.OpenShiftServerUtils.findServerForService(Unknown Source)
	at org.jboss.tools.openshift.internal.ui.handler.ServerAdapterHandler.openOrCreateServerAdapter(ServerAdapterHandler.java:113)
	at org.jboss.tools.openshift.internal.ui.handler.ServerAdapterHandler.getOpenShiftServer(ServerAdapterHandler.java:81)
	at org.jboss.tools.openshift.internal.ui.handler.ServerAdapterHandler.execute(ServerAdapterHandler.java:56)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:295)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:55)
	... 40 more
```

**Pull request merged**